File: bot.py
"""
bot.py – entry point for the Discord / Trello bot.

Usage
-----
    python bot.py

Environment variables (see .env.example)
-----------------------------------------
    DISCORD_TOKEN       – Discord bot token
    TRELLO_API_KEY      – Trello API key
    TRELLO_TOKEN        – Trello user token
    TRELLO_BOARD_ID     – Target Trello board ID
    DISCORD_GUILD_ID    – (optional) guild ID for fast slash-command sync
"""

import asyncio
import logging

import discord
from discord.ext import commands

import config
from cogs.trello_commands import TrelloCommands
from utils.trello_client import TrelloClient

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(name)s: %(message)s",
)
logger = logging.getLogger(__name__)


class BotGeral(commands.Bot):
    def __init__(self) -> None:
        intents = discord.Intents.default()
        super().__init__(command_prefix="!", intents=intents)
        self._guild_id = config.DISCORD_GUILD_ID

    async def setup_hook(self) -> None:
        trello = TrelloClient(
            api_key=config.TRELLO_API_KEY,
            token=config.TRELLO_TOKEN,
            board_id=config.TRELLO_BOARD_ID,
        )
        self.tree.add_command(TrelloCommands(trello))

        if self._guild_id:
            guild = discord.Object(id=self._guild_id)
            self.tree.copy_global_to(guild=guild)
            synced = await self.tree.sync(guild=guild)
            logger.info("Synced %d command(s) to guild %s", len(synced), self._guild_id)
        else:
            synced = await self.tree.sync()
            logger.info("Synced %d global command(s)", len(synced))

    async def on_ready(self) -> None:
        logger.info("Logged in as %s (ID: %s)", self.user, self.user.id if self.user else "?")
        await self.change_presence(
            activity=discord.Activity(
                type=discord.ActivityType.watching,
                name="o quadro Trello 📋",
            )
        )


async def main() -> None:
    async with BotGeral() as bot:
        await bot.start(config.DISCORD_TOKEN)


if __name__ == "__main__":
    asyncio.run(main())

File: src/bot.js
/**
 * src/bot.js – Discord bot entrypoint.
 *
 * Usage:
 *   node src/bot.js   (or npm start)
 *
 * Required environment variables (see .env.example):
 *   DISCORD_TOKEN       Discord bot token
 *   TRELLO_API_KEY      Trello API key
 *   TRELLO_TOKEN        Trello user token
 *   TRELLO_BOARD_ID     Target Trello board ID
 */

require('dotenv').config();
const { Client, GatewayIntentBits, ActivityType } = require('discord.js');
const { TrelloClient } = require('./trelloClient');
const trelloCommand = require('./commands/trello');

// ── Validation ────────────────────────────────────────────────────────────────

const required = ['DISCORD_TOKEN', 'TRELLO_API_KEY', 'TRELLO_TOKEN', 'TRELLO_BOARD_ID'];
for (const key of required) {
  if (!process.env[key]) {
    console.error(`Missing required environment variable: ${key}`);
    process.exit(1);
  }
}

// ── Discord client ────────────────────────────────────────────────────────────

const client = new Client({ intents: [GatewayIntentBits.Guilds] });

const trello = new TrelloClient(
  process.env.TRELLO_API_KEY,
  process.env.TRELLO_TOKEN,
  process.env.TRELLO_BOARD_ID,
);

// ── Event: ready ──────────────────────────────────────────────────────────────

client.once('ready', (c) => {
  console.log(`✅ Logged in as ${c.user.tag} (ID: ${c.user.id})`);
  c.user.setActivity('o quadro Trello 📋', { type: ActivityType.Watching });
});

// ── Event: interactionCreate ──────────────────────────────────────────────────

client.on('interactionCreate', async (interaction) => {
  if (!interaction.isChatInputCommand()) return;
  if (interaction.commandName !== 'trello') return;

  try {
    await trelloCommand.execute(interaction, trello);
  } catch (err) {
    console.error('Unhandled error in trello command:', err);
    const msg = '❌ Ocorreu um erro inesperado. Tente novamente mais tarde.';
    if (interaction.replied || interaction.deferred) {
      await interaction.followUp({ content: msg, ephemeral: true });
    } else {
      await interaction.reply({ content: msg, ephemeral: true });
    }
  }
});

// ── Start ─────────────────────────────────────────────────────────────────────

client.login(process.env.DISCORD_TOKEN);
